### ISSUE TITLE
:(tab)open to use bookmark keyword, with keyword completion

### ISSUE THREAD

**Issue by MilanV** [tridactyl/tridactyl#1312]
For "`http://...../..../..%s`" bookmarks with a keyword set, `:open keyword text` should open that URL with "`%s`" replaced with "`text`". Also, while typing first word after `:(tab)open`, the completion should offer bookmark keywords.

Note: I know it was believed FF API does not allow the extension to see such keywords, there are some closed issues from past about this topic. But [Second Search extension](https://addons.mozilla.org/en-US/firefox/addon/second-search/) knows to do that. Can we inspire there?

**Comment by bovine3dom** [comment]
I'll add it to my list of things to look into. Their repository is here: https://github.com/piroor/secondsearch.

**Comment by glacambre** [comment]
We can't access the keyword and secondsearch doesn't access the keyword either. It builds the list of search urls from the bookmarks that contain "%s" and then uses the title of the bookmarks in order to display the search engine name (see their [background.js](https://github.com/piroor/secondsearch/blob/master/webextensions/background/background.js) :/

**Comment by glacambre** [comment]
Yeah, importing searchurls from bookmarks has been mentioned a bunch of times. The problem is that you can't guess what keywords to give them. Can't use the title because it might contain spaces.

**Comment by MilanV** [comment]
I must say I haven't had time to check Second Search sources yet but it also allows you marking some input field with a keyword - and add it to the list of keyworded bookmarks. Is it something different than original keywod property of FF bookmark? Something extension-specific?

**Comment by glacambre** [comment]
> it also allows you marking some input field with a keyword

That's a default Firefox feature. You can already do that without SecondSearch being installed.

**Comment by MilanV** [comment]
> That's a default Firefox feature. You can already do that without SecondSearch being installed.

Yes but SecondSearch can remember such keywords - that was my idea. But now, I see they do nothing special, just let FF creating the new bookmark and they get it on their list of "keywords" automatically because of the "%s" substring in its URL. Neither SecondSearch can track what keyword you filled in for that bookmark - that's why they list them by title only :(

So, unfortunately, my hope died, I do not see anything we can get from SecondSearch about bookmark keywords. We have to hope in FF API enhancement. Or do you see anything how we can help meanwhile, until the FF API is extended (if ever)?

**Comment by glacambre** [comment]
You could manually turn your search keywords into browser search engines where possible since Tridactyl now supports dispatching search queries with them. Another possibility would be to use Tridactyl's searchurls (`:help searchurls`), they're quite similar and even slightly more powerful than search keywords.

If you want to go with searchurls, you could try running the following snippet in Tridactyl's command line in order to import your search keywords:
```
s tri.browserBg.bookmarks.search({query:""}).then(bookmarks => bookmarks.filter(bmark => bmark.url && bmark.url.includes("%s")).forEach(bmark => { let keyword = prompt(bmark.title + " " + bmark.url) ; if (keyword && keyword.length > 0) { tri.config.set("searchurls", keyword, bmark.url) } }))
```
It will prompt you for a keyword for each url you have that contain "%s" and create the corresponding searchurl.
